docbrown/src/db/view_api/time.rs

```diff
@@ -41,55 +42,238 @@ pub trait TimeOps {
     /// using an expanding window.
     ///
     /// An expanding window is a window that grows by `step` size at each iteration.
-    fn expanding(&self, step: u64, start: Option<i64>, end: Option<i64>) -> WindowSet<Self>
+    fn expanding<I>(&self, step: I) -> Result<WindowSet<Self>, ParseTimeError>
     where
-        Self: Sized + Clone,
+        Self: Sized + Clone + 'static, // TODO: is this fine???
+        I: TryInto<Interval, Error = ParseTimeError>,
     {
-        self.through_perspectives(Perspective::expanding(step, start, end))
+        let parent = self.clone();
+        match (self.start(), self.end()) {
+            (Some(start), Some(end)) => {
+                let step: Interval = step.try_into()?;
+
+                Ok(WindowSet::new(parent, start, end, step, None))
+            }
+            _ => Ok(WindowSet::empty(parent)),
+        }
     }
 
     /// Creates a `WindowSet` with the given `window` size and optional `step`, `start` and `end` times,
     /// using a rolling window.
     ///
     /// A rolling window is a window that moves forward by `step` size at each iteration.
-    fn rolling(
-        &self,
-        window: u64,
-        step: Option<u64>,
-        start: Option<i64>,
-        end: Option<i64>,
-    ) -> WindowSet<Self>
+    fn rolling<I>(&self, window: I, step: Option<I>) -> Result<WindowSet<Self>, ParseTimeError>
     where
-        Self: Sized + Clone,
+        Self: Sized + Clone + 'static, // TODO: is this fine???
+        I: TryInto<Interval, Error = ParseTimeError>,
     {
-        self.through_perspectives(Perspective::rolling(window, step, start, end))
+        let parent = self.clone();
+        match (self.start(), self.end()) {
+            (Some(start), Some(end)) => {
+                let window: Interval = window.try_into()?;
+                let step: Interval = match step {
+                    Some(step) => step.try_into()?,
+                    None => window,
+                };
+                Ok(WindowSet::new(parent, start, end, step, Some(window)))
+            }
+            _ => Ok(WindowSet::empty(parent)),
+        }
     }
 
-    /// Creates a `WindowSet` from a set of perspectives
-    fn through_perspectives(&self, perspectives: PerspectiveSet) -> WindowSet<Self>
-    where
-        Self: Sized + Clone,
-    {
-        let iter = match (self.start(), self.end()) {
-            (Some(start), Some(end)) => perspectives.build_iter(start..end),
-            _ => PerspectiveIterator::empty(),
+    // TODO pub fn weeks(n), days(n), hours(n), minutes(n), seconds(n), millis(n)
+}
+
+pub struct WindowSet<T: TimeOps> {
+    view: T,
+    cursor: i64,
+    end: i64,
+    step: Interval,
+    window: Option<Interval>,
+}
+
+impl<T: TimeOps> WindowSet<T> {
+    fn new(
+        view: T,
+        timeline_start: i64,
+        end: i64,
+        step: Interval,
+        window: Option<Interval>,
+    ) -> Self {
+        let cursor_start = if step.epoch_alignment {
```

Comment:
Have a double check over this.

Reply:
add_vertex(1682005273,1,..) <-- seconds

g.at("21/04/2021") <- assumed milliseconds